### PR TITLE
UNDERTOW-1487: SimpleObjectPool.close may be invoked multiple times

### PR DESCRIPTION
**Pull request by cakofony**

### CODE REVIEW

File: core/src/test/java/io/undertow/util/SimpleObjectPoolTestCase.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
+package io.undertow.util;
```
Comment (fl4via) [inline core/src/test/java/io/undertow/util/SimpleObjectPoolTestCase.java]: We are missing the license header and a javadoc with @author

Comment (cakofony) [reply core/src/test/java/io/undertow/util/SimpleObjectPoolTestCase.java]: Updated

**Comment by undertow-pull-request** [comment]
**Windows** [Build 2674](https://ci.wildfly.org/viewLog.html?buildId=140375&buildTypeId=UNDERTOW_MasterWindows) outcome was **FAILURE** using a merge of 0d108a721a031763d797f303fbfac3da2934ed88
Summary: Tests failed: 1 (1 new), passed: 3375, ignored: 709 Build time: 00:49:06
### Failed tests
```
io.undertow.server.handlers.ReceiverTestCase.testAsyncReceiveWholeBytesFailed[proxy][http2]: java.lang.AssertionError
	at io.undertow.server.handlers.ReceiverTestCase.testAsyncReceiveWholeBytesFailed(ReceiverTestCase.java:194)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:524)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:104)
	at io.undertow.testutils.DefaultServer.run(DefaultServer.java:297)
------- Stderr: -------
20:46:52,305 DEBUG (XNIO-1 I/O-3) [io.undertow.server.handlers.proxy] <ProxyHandler.java:549> Sending request ClientRequest{path='/fullbytes', method=POST, protocol=HTTP/1.1} to target localhost/127.0.0.1:8888 for exchange HttpServerExchange{ POST /fullbytes request {Content-Length=[130000100], Host=[localhost]} response {}}
20:46:52,306 DEBUG (XNIO-1 I/O-3) [io.undertow.server.handlers.proxy] <ProxyHandler.java:566> Sent request ClientRequest{path='/fullbytes', method=POST, protocol=HTTP/1.1} to target 127.0.0.1 for exchange HttpServerExchange{ POST /fullbytes request {Content-Length=[130000100], Host=[localhost]} response {}}
20:46:52,377 ERROR (XNIO-1 I/O-3) [io.undertow.proxy] <ProxyHandler.java:662> UT005028: Proxy request to /fullbytes failed: java.nio.channels.ClosedChannelException
	at io.undertow.client.http2.Http2ClientConnection$1.handleEvent(Http2ClientConnection.java:107)
	at io.undertow.client.http2.Http2ClientConnection$1.handleEvent(Http2ClientConnection.java:99)
	at org.xnio.ChannelListeners.invokeChannelListener(ChannelListeners.java:92)
	at io.undertow.server.protocol.framed.AbstractFramedChannel$FrameCloseListener.handleEvent(AbstractFramedChannel.java:1061)
	at io.undertow.server.protocol.framed.AbstractFramedChannel$FrameCloseListener.handleEvent(AbstractFramedChannel.java:975)
	at org.xnio.ChannelListeners.invokeChannelListener(ChannelListeners.java:92)
	at io.undertow.protocols.ssl.SslConduit$SslReadReadyHandler.terminated(SslConduit.java:1214)
	at org.xnio.nio.NioSocketConduit.readTerminated(NioSocketConduit.java:360)
	at org.xnio.nio.NioSocketStreamConnection.notifyReadClosed(NioSocketStreamConnection.java:148)
	at org.xnio.Connection.close(Connection.java:139)
	at org.xnio.IoUtils.safeClose(IoUtils.java:134)
	at io.undertow.protocols.ssl.SslConduit.closed(SslConduit.java:1046)
	at io.undertow.protocols.ssl.SslConduit.close(SslConduit.java:1117)
	at io.undertow.protocols.ssl.UndertowSslConnection.closeAction(UndertowSslConnection.java:149)
	at org.xnio.Connection.close(Connection.java:132)
	at org.xnio.IoUtils.safeClose(IoUtils.java:134)
	at io.undertow.server.protocol.framed.AbstractFramedChannel.close(AbstractFramedChannel.java:816)
	at org.xnio.IoUtils.safeClose(IoUtils.java:134)
	at io.undertow.protocols.http2.Http2Channel.sendGoAway(Http2Channel.java:832)
	at io.undertow.protocols.http2.Http2Channel.sendGoAway(Http2Channel.java:809)
	at io.undertow.client.http2.Http2ClientConnection.close(Http2ClientConnection.java:303)
	at org.xnio.IoUtils.safeClose(IoUtils.java:134)
	at io.undertow.server.handlers.proxy.ProxyHandler$IoExceptionHandler.handleException(ProxyHandler.java:790)
	at org.xnio.ChannelListeners.invokeChannelExceptionHandler(ChannelListeners.java:126)
	at io.undertow.util.Transfer$TransferListener.handleEvent(Transfer.java:218)
	at org.xnio.ChannelListeners.invokeChannelListener(ChannelListeners.java:92)
	at io.undertow.server.protocol.framed.AbstractFramedStreamSinkChannel$1.run(AbstractFramedStreamSinkChannel.java:214)
	at io.undertow.server.protocol.framed.AbstractFramedChannel$FrameReadListener.handleEvent(AbstractFramedChannel.java:935)
	at io.undertow.server.protocol.framed.AbstractFramedChannel$FrameReadListener.handleEvent(AbstractFramedChannel.java:929)
	at org.xnio.ChannelListeners.invokeChannelListener(ChannelListeners.java:92)
	at org.xnio.conduits.ReadReadyHandler$ChannelListenerHandler.readReady(ReadReadyHandler.java:66)
	at io.undertow.protocols.ssl.SslConduit$SslReadReadyHandler.readReady(SslConduit.java:1175)
	at io.undertow.protocols.ssl.SslConduit$1.run(SslConduit.java:168)
	at org.xnio.nio.WorkerThread.safeRun(WorkerThread.java:582)
	at org.xnio.nio.WorkerThread.run(WorkerThread.java:466)

20:46:52,378 DEBUG (XNIO-1 I/O-3) [io.undertow.request.error-response] <DelegatingBasicLogger.java:424> Setting error code 503 for exchange HttpServerExchange{ POST /fullbytes request {Content-Length=[130000100], Host=[localhost]} response {}}: java.lang.RuntimeException
	at io.undertow.server.HttpServerExchange.setStatusCode(HttpServerExchange.java:1410)
	at io.undertow.server.handlers.proxy.ProxyHandler.handleFailure(ProxyHandler.java:668)
	at io.undertow.server.handlers.proxy.ProxyHandler$ResponseCallback.failed(ProxyHandler.java:727)
	at io.undertow.client.http2.Http2ClientExchange.failed(Http2ClientExchange.java:125)
	at io.undertow.client.http2.Http2ClientConnection$1.handleEvent(Http2ClientConnection.java:107)


```

**Comment by fl4via** [comment]
Retest this please

**Comment by undertow-pull-request** [comment]
**Windows** [Build 2678](https://ci.wildfly.org/viewLog.html?buildId=140390&buildTypeId=UNDERTOW_MasterWindows) outcome was **FAILURE** using a merge of 0d108a721a031763d797f303fbfac3da2934ed88
Summary: Tests failed: 1, passed: 3379, ignored: 709 Build time: 00:46:17
### Failed tests
```
io.undertow.server.handlers.ReceiverTestCase.testAsyncReceiveWholeBytesFailed[proxy][http2]: java.lang.AssertionError
	at io.undertow.server.handlers.ReceiverTestCase.testAsyncReceiveWholeBytesFailed(ReceiverTestCase.java:194)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:524)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:104)
	at io.undertow.testutils.DefaultServer.run(DefaultServer.java:297)
------- Stderr: -------
21:39:31,839 DEBUG (XNIO-1 I/O-8) [io.undertow.server.handlers.proxy] <ProxyHandler.java:549> Sending request ClientRequest{path='/fullbytes', method=POST, protocol=HTTP/1.1} to target localhost/127.0.0.1:8888 for exchange HttpServerExchange{ POST /fullbytes request {Content-Length=[130000100], Host=[localhost]} response {}}
21:39:31,839 DEBUG (XNIO-1 I/O-8) [io.undertow.server.handlers.proxy] <ProxyHandler.java:566> Sent request ClientRequest{path='/fullbytes', method=POST, protocol=HTTP/1.1} to target 127.0.0.1 for exchange HttpServerExchange{ POST /fullbytes request {Content-Length=[130000100], Host=[localhost]} response {}}
21:39:31,889 ERROR (XNIO-1 I/O-8) [io.undertow.proxy] <ProxyHandler.java:662> UT005028: Proxy request to /fullbytes failed: java.nio.channels.ClosedChannelException
	at io.undertow.client.http2.Http2ClientConnection$1.handleEvent(Http2ClientConnection.java:107)
	at io.undertow.client.http2.Http2ClientConnection$1.handleEvent(Http2ClientConnection.java:99)
	at org.xnio.ChannelListeners.invokeChannelListener(ChannelListeners.java:92)
	at io.undertow.server.protocol.framed.AbstractFramedChannel$FrameCloseListener.handleEvent(AbstractFramedChannel.java:1061)
	at io.undertow.server.protocol.framed.AbstractFramedChannel$FrameCloseListener.handleEvent(AbstractFramedChannel.java:975)
	at org.xnio.ChannelListeners.invokeChannelListener(ChannelListeners.java:92)
	at io.undertow.protocols.ssl.SslConduit$SslReadReadyHandler.terminated(SslConduit.java:1214)
	at org.xnio.nio.NioSocketConduit.readTerminated(NioSocketConduit.java:360)
	at org.xnio.nio.NioSocketStreamConnection.notifyReadClosed(NioSocketStreamConnection.java:148)
	at org.xnio.Connection.close(Connection.java:139)
	at org.xnio.IoUtils.safeClose(IoUtils.java:134)
	at io.undertow.protocols.ssl.SslConduit.closed(SslConduit.java:1046)
	at io.undertow.protocols.ssl.SslConduit.close(SslConduit.java:1117)
	at io.undertow.protocols.ssl.UndertowSslConnection.closeAction(UndertowSslConnection.java:149)
	at org.xnio.Connection.close(Connection.java:132)
	at org.xnio.IoUtils.safeClose(IoUtils.java:134)
	at io.undertow.server.protocol.framed.AbstractFramedChannel.close(AbstractFramedChannel.java:816)
	at org.xnio.IoUtils.safeClose(IoUtils.java:134)
	at io.undertow.protocols.http2.Http2Channel.sendGoAway(Http2Channel.java:832)
	at io.undertow.protocols.http2.Http2Channel.sendGoAway(Http2Channel.java:809)
	at io.undertow.client.http2.Http2ClientConnection.close(Http2ClientConnection.java:303)
	at org.xnio.IoUtils.safeClose(IoUtils.java:134)
	at io.undertow.server.handlers.proxy.ProxyHandler$IoExceptionHandler.handleException(ProxyHandler.java:790)
	at org.xnio.ChannelListeners.invokeChannelExceptionHandler(ChannelListeners.java:126)
	at io.undertow.util.Transfer$TransferListener.handleEvent(Transfer.java:218)
	at org.xnio.ChannelListeners.invokeChannelListener(ChannelListeners.java:92)
	at io.undertow.server.protocol.framed.AbstractFramedStreamSinkChannel$1.run(AbstractFramedStreamSinkChannel.java:214)
	at io.undertow.server.protocol.framed.AbstractFramedChannel$1.run(AbstractFramedChannel.java:137)
	at org.xnio.nio.WorkerThread.safeRun(WorkerThread.java:582)
	at org.xnio.nio.WorkerThread.run(WorkerThread.java:466)

21:39:31,890 DEBUG (XNIO-1 I/O-2) [io.undertow.request.io] <AbstractFramedChannel.java:840> Marking reads broken on channel Http2Channel peer /127.0.0.1:62641 local 0.0.0.0/0.0.0.0:8888[ No Receiver [] -- [] -- []]: java.nio.channels.ClosedChannelException
	at io.undertow.server.protocol.framed.AbstractFramedChannel.markReadsBroken(AbstractFramedChannel.java:840)
	at io.undertow.server.protocol.framed.AbstractFramedChannel.receive(AbstractFramedChannel.java:476)
	at io.undertow.server.protocol.http2.Http2ReceiveListener.handleEvent(Http2ReceiveListener.java:118)
	at io.undertow.server.protocol.http2.Http2ReceiveListener.handleEvent(Http2ReceiveListener.java:67)
	at org.xnio.ChannelListeners.invokeChannelListener(ChannelListeners.java:92)
	at io.undertow.server.protocol.framed.AbstractFramedChannel$FrameReadListener.handleEvent(AbstractFramedChannel.java:948)
	at io.undertow.server.protocol.framed.AbstractFramedChannel$FrameReadListener.handleEvent(AbstractFramedChannel.java:929)
	at org.xnio.ChannelListeners.invokeChannelListener(ChannelListeners.java:92)
	at org.xnio.conduits.ReadReadyHandler$ChannelListenerHandler.readReady(ReadReadyHandler.java:66)
	at io.undertow.protocols.ssl.SslConduit$SslReadReadyHandler.readReady(SslConduit.java:1175)


```

**Comment by fl4via** [comment]
Retest this please

**Pull request merged**